bpf/go_sql.c

```diff
@@ -0,0 +1,104 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "common.h"
+#include "bpf_helpers.h"
+#include "bpf_builtins.h"
+#include "go_common.h"
+#include "bpf_dbg.h"
+#include <stdbool.h>
+
+#define MAX_QUERY_SIZE 100
+
+#define BUFFER_QUERY_ARG 0
+
+struct sql_request_t {
+    u64 start_time;
+#if BUFFER_QUERY_ARG
+    char query[MAX_QUERY_SIZE];
+#else
+    char *query_ptr;
+    u8    query_len;
+#endif
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, void*);
+	__type(value, struct sql_request_t);
+	__uint(max_entries, MAX_CONCURRENT_REQUESTS);
+} sql_events SEC(".maps");
+
+// Injected in init
+volatile const bool should_include_db_statement;
+
+// This instrumentation attaches uprobe to the following function:
+// func (db *DB) queryDC(ctx, txctx context.Context, dc *driverConn, releaseConn func(error), query string, args []any)
+SEC("uprobe/queryDC")
+int uprobe_queryDC(struct pt_regs *ctx) {
+    bpf_dbg_printk("=== uprobe/queryDC === ");
+    void *goroutine_addr = GOROUTINE_PTR(ctx);
+    bpf_dbg_printk("goroutine_addr %lx", goroutine_addr);
+
+    struct sql_request_t sql_request = {0};
+    sql_request.start_time = bpf_ktime_get_ns();
+
+    if (should_include_db_statement) {
```

Comment:
I'm not sure what happened to my comment here. I believe the more optimal approach is to do all the reading at the return of the queryDC. There are couple of reasons for that:

- We don't need an additional buffer, we can read directly into the trace event buffer that's allocated from the ringbuf.
- We can reuse our function invocation data structure, e.g. no need for another struct to be maintaned.

This approach is also different than OTel and we don't have to carry the copyright notice for code that's mostly written by Grafana. If we did this I don't think we need to copy any code from Otel.

Reply:
OK I've refactored all otel code out, and made it more consistent with the other probes by copying the full set of registers into an ongoing transaction map and using a few of them in the return probe.